src/fairseq2/utils/rng.py

```diff
@@ -40,3 +26,90 @@ def use_deterministic(value: bool, warn_only: bool = False) -> None:
     torch.backends.cudnn.benchmark = not value
 
     torch.use_deterministic_algorithms(value, warn_only=warn_only)
+
+
+class GeneratorBag:
+    """Holds a collection of random number generators."""
+
+    generators: List[Generator]
+
+    def __init__(self, *generators: Generator) -> None:
+        """
+        :param generators:
+            The generators to hold.
+        """
+        self.generators = list(generators)
+
+    @staticmethod
+    def from_device_defaults(*devices: Device) -> GeneratorBag:
+        """Create a :class:`GeneratorBag` instance holding the default random
+        number generators of ``devices``."""
+        generators = []
+
+        for device in set(devices):
+            if device.type == "cpu":
+                generators.append(torch.default_generator)
+            elif device.type == "cuda":
+                # Ensure that the default CUDA generators are initialized.
+                torch.cuda.init()
+
+                idx = device.index
+                if idx is None:
+                    idx = torch.cuda.current_device()
+
+                generators.append(torch.cuda.default_generators[idx])
+            else:
+                raise ValueError(
+                    f"`GeneratorBag` supports only CPU and CUDA devices, but a {device.type.upper()} is specified."
+                )
+
+        return GeneratorBag(*generators)
+
+    def seed(self) -> None:
+        """Set the seed of the random number generators to a random number."""
+        if not self.generators:
+            return
+
+        self.generators[0].seed()
+
+        random_seed = self.generators[0].initial_seed()
+
+        for g in self.generators[1:]:
+            g.manual_seed(random_seed)
+
+    def manual_seed(self, seed: int) -> None:
+        """Set the seed of the random number generators."""
+        if seed >= 1 << 32:
```

Comment:
Does this condition check if seed is negative ?

Reply:
Not really. Let me add that check as well.